File: dagster_poc/resources.py
import boto3
from dagster import ConfigurableResource


class ECSResource(ConfigurableResource):
    """Resource para interactuar con AWS ECS/Fargate."""

    region_name: str = "us-east-1"
    cluster_name: str = "default"
    subnets: list[str] = []
    security_groups: list[str] = []
    log_group_name: str = ""

    # Task definitions por tamaño
    task_definition_small: str = ""
    task_definition_medium: str = ""
    task_definition_large: str = ""
    task_definition_xlarge: str = ""

    def get_client(self):
        return boto3.client("ecs", region_name=self.region_name)

    def get_task_definition(self, task_size: str) -> str:
        """Get task definition ARN based on size."""
        mapping = {
            "small": self.task_definition_small,
            "medium": self.task_definition_medium,
            "large": self.task_definition_large,
            "xlarge": self.task_definition_xlarge,
        }
        return mapping.get(task_size, self.task_definition_small)

    def get_network_config(self, assign_public_ip: bool = True) -> dict:
        """Get network configuration for Fargate task."""
        return {
            "awsvpcConfiguration": {
                "subnets": self.subnets,
                "securityGroups": self.security_groups,
                "assignPublicIp": "ENABLED" if assign_public_ip else "DISABLED",
            }
        }


class SQSResource(ConfigurableResource):
    """Resource para interactuar con AWS SQS."""

    region_name: str = "us-east-1"
    queue_url: str = ""

    def get_client(self):
        return boto3.client("sqs", region_name=self.region_name)

    def receive_messages(self, max_messages: int = 10, wait_time: int = 5) -> list[dict]:
        """Recibe mensajes de la cola."""
        client = self.get_client()
        response = client.receive_message(
            QueueUrl=self.queue_url,
            MaxNumberOfMessages=max_messages,
            WaitTimeSeconds=wait_time,
        )
        return response.get("Messages", [])

    def delete_message(self, receipt_handle: str) -> None:
        """Elimina un mensaje de la cola."""
        client = self.get_client()
        client.delete_message(QueueUrl=self.queue_url, ReceiptHandle=receipt_handle)


class S3Resource(ConfigurableResource):
    """Resource para interactuar con AWS S3."""

    region_name: str = "us-east-1"

    def get_client(self):
        return boto3.client("s3", region_name=self.region_name)

    def get_file_size(self, bucket: str, key: str) -> int:
        """Get file size in bytes."""
        client = self.get_client()
        response = client.head_object(Bucket=bucket, Key=key)
        return response.get("ContentLength", 0)

    def get_recommended_task_size(self, file_size_bytes: int) -> str:
        """Determine optimal task size based on file size."""
        size_mb = file_size_bytes / (1024 * 1024)

        if size_mb < 50:
            return "small"
        elif size_mb < 200:
            return "medium"
        elif size_mb < 500:
            return "large"
        else:
            return "xlarge"

File: dagster_ts/src/resources.ts
/**
 * AWS Resources - TypeScript equivalent of dagster_poc/resources.py
 */

import { ECSClient } from "@aws-sdk/client-ecs";
import {
  SQSClient,
  ReceiveMessageCommand,
  DeleteMessageCommand,
} from "@aws-sdk/client-sqs";
import { S3Client, HeadObjectCommand } from "@aws-sdk/client-s3";

// ─── ECS Resource ───────────────────────────────────────────────

export interface ECSResourceConfig {
  regionName?: string;
  clusterName: string;
  subnets: string[];
  securityGroups: string[];
  logGroupName: string;
  dynamoTableName: string;
  taskDefinitionSmall: string;
  taskDefinitionMedium: string;
  taskDefinitionLarge: string;
  taskDefinitionXlarge: string;
}

export class ECSResource {
  readonly regionName: string;
  readonly clusterName: string;
  readonly subnets: string[];
  readonly securityGroups: string[];
  readonly logGroupName: string;
  readonly dynamoTableName: string;
  readonly taskDefinitionSmall: string;
  readonly taskDefinitionMedium: string;
  readonly taskDefinitionLarge: string;
  readonly taskDefinitionXlarge: string;

  private client: ECSClient;

  constructor(config: ECSResourceConfig) {
    this.regionName = config.regionName ?? "us-east-1";
    this.clusterName = config.clusterName;
    this.subnets = config.subnets;
    this.securityGroups = config.securityGroups;
    this.logGroupName = config.logGroupName;
    this.dynamoTableName = config.dynamoTableName;
    this.taskDefinitionSmall = config.taskDefinitionSmall;
    this.taskDefinitionMedium = config.taskDefinitionMedium;
    this.taskDefinitionLarge = config.taskDefinitionLarge;
    this.taskDefinitionXlarge = config.taskDefinitionXlarge;
    this.client = new ECSClient({ region: this.regionName });
  }

  getClient(): ECSClient {
    return this.client;
  }

  getTaskDefinition(taskSize: string): string {
    const mapping: Record<string, string> = {
      small: this.taskDefinitionSmall,
      medium: this.taskDefinitionMedium,
      large: this.taskDefinitionLarge,
      xlarge: this.taskDefinitionXlarge,
    };
    return mapping[taskSize] ?? this.taskDefinitionSmall;
  }

  getNetworkConfig(assignPublicIp = true) {
    return {
      awsvpcConfiguration: {
        subnets: this.subnets,
        securityGroups: this.securityGroups,
        assignPublicIp: assignPublicIp ? ("ENABLED" as const) : ("DISABLED" as const),
      },
    };
  }
}

// ─── SQS Resource ───────────────────────────────────────────────

export interface SQSResourceConfig {
  regionName?: string;
  queueUrl: string;
}

export interface SQSMessage {
  MessageId?: string;
  ReceiptHandle?: string;
  Body?: string;
}

export class SQSResource {
  readonly regionName: string;
  readonly queueUrl: string;
  private client: SQSClient;

  constructor(config: SQSResourceConfig) {
    this.regionName = config.regionName ?? "us-east-1";
    this.queueUrl = config.queueUrl;
    this.client = new SQSClient({ region: this.regionName });
  }

  getClient(): SQSClient {
    return this.client;
  }

  async receiveMessages(maxMessages = 10, waitTime = 5): Promise<SQSMessage[]> {
    const command = new ReceiveMessageCommand({
      QueueUrl: this.queueUrl,
      MaxNumberOfMessages: maxMessages,
      WaitTimeSeconds: waitTime,
    });
    const response = await this.client.send(command);
    return (response.Messages ?? []) as SQSMessage[];
  }

  async deleteMessage(receiptHandle: string): Promise<void> {
    const command = new DeleteMessageCommand({
      QueueUrl: this.queueUrl,
      ReceiptHandle: receiptHandle,
    });
    await this.client.send(command);
  }
}

// ─── Lambda Resource ────────────────────────────────────────────

export interface LambdaResourceConfig {
  regionName?: string;
  functionName: string;
  logGroupName: string;
  dynamoTableName: string;
}

export class LambdaResource {
  readonly regionName: string;
  readonly functionName: string;
  readonly logGroupName: string;
  readonly dynamoTableName: string;

  constructor(config: LambdaResourceConfig) {
    this.regionName = config.regionName ?? "us-east-1";
    this.functionName = config.functionName;
    this.logGroupName = config.logGroupName;
    this.dynamoTableName = config.dynamoTableName;
  }
}

// ─── S3 Resource ────────────────────────────────────────────────

export interface S3ResourceConfig {
  regionName?: string;
}

export class S3Resource {
  readonly regionName: string;
  private client: S3Client;

  constructor(config: S3ResourceConfig = {}) {
    this.regionName = config.regionName ?? "us-east-1";
    this.client = new S3Client({ region: this.regionName });
  }

  getClient(): S3Client {
    return this.client;
  }

  async getFileSize(bucket: string, key: string): Promise<number> {
    const command = new HeadObjectCommand({ Bucket: bucket, Key: key });
    const response = await this.client.send(command);
    return response.ContentLength ?? 0;
  }

  getRecommendedTaskSize(fileSizeBytes: number): string {
    const sizeMb = fileSizeBytes / (1024 * 1024);

    if (sizeMb < 50) return "lambda";
    if (sizeMb < 200) return "medium";
    if (sizeMb < 500) return "large";
    return "xlarge";
  }
}
